File: VerticalTree.py
from Tree import Tree

import math
from PySide.QtCore import QPointF, Qt, QPoint, QRect
from PySide.QtGui import QPixmap, QPainter

class VerticalTree(Tree):
    _max = 0
    _min = 0
    _directLink = False
    
    def __init__(self, parent=None):
        super(VerticalTree, self).__init__( parent )
        
                
    def setDirectConnection(self, direct = True):
        self._directLink = direct
     
     
    def setScrollBarValues(self):
        dw = max( [0, (self.realSize.width() - self.width())] )
        dh = max( [0, (self.realSize.height() - self.height()) ] )
        self.horizontalScrollBar().setRange( -dw/2, dw/2 )
        self.verticalScrollBar().setRange( 0, dh )
        
        if self.model() == None:
            return None
        
        index = self.model().index(0,0)
        if index in self.itemPos:
            p = self.itemPos[ index ]
            self.itemOffset = QPointF( self.width() / 2 - p.x(), 0 )


    def scan(self, index, left, depth):
        rows = self.model().rowCount( index )
        if not index.isValid():
            return 0
        elif rows == 0:
            self.setX( index, left )
            self.setY(index, depth + 1)
            return (left + 1, 1)
        
        childDepth = 0
        for r in range( 0, self.model().rowCount( index ) ):
            child = index.child( r, 0 )
            (left, d) = self.scan( child, left, depth + 1 )
            childDepth = max( childDepth, d )
        
        left = self.itemTreePos[ self.model().index(0, index.column(), index) ].x()
        right = self.itemTreePos[ self.model().index( rows - 1, index.column(), index ) ].x()
        if rows >= 2:
            if rows % 2 == 1:
                r = math.floor( rows/2 ) + 1
                v = self.itemTreePos[ self.model().index( r - 1, index.column(), index ) ].x()
                self.setX( index, v )
            else:
                self.setX( index, (right + left) / 2 )
        else:
            self.setX( index, left )
            
        self.setY(index, depth)
        return ( right + 1 , childDepth + 1 ) 
            
        
    def positionsInTree(self):
        self.itemTreePos = {}
        (self.left, self.depth) = self.scan( self.model().index(0,0) , 0, 0)
            
        self.positionsInView()
        
    def positionsInView(self):
        self.realSize.setWidth( ( self.left + 1 ) * ( self.xDistance + self.rect.width() ) + 40 );
        self.realSize.setHeight( self.depth * ( self.yDistance + self.rect.height() ) + 40 )
        
        self.setScrollBarValues()
        
        for index in self.itemTreePos.keys():
            p = self.itemTreePos[ index ]
            self.itemPos[index] = QPointF( ( 1 + p.x() ) * ( self.xDistance + self.rect.width() ), p.y() * ( self.yDistance + self.rect.height() ) )
    
    
    def paintConnectionsFor(self, painter, index, offset ):
        painter.save()
        painter.setPen( self._connectionPen )
        if self._directLink == True:
            super(VerticalTree, self).paintConnectionsFor( painter, index, offset )
        else:
            rows = self.model().rowCount( index )
            r = self.itemRect(index).translated( offset.x(), offset.y() )
            if rows > 1:
                child1 = self.itemRect( self.model().index( 0, 0, index ) ).translated( offset.x(), offset.y() )
                child2 = self.itemRect( self.model().index( rows - 1, 0, index ) ).translated( offset.x(), offset.y() )
                p1 = QPointF( child1.center().x(), child1.top() - self.yDistance / 2 )
                p2 = QPointF( child2.center().x(), child2.top() - self.yDistance / 2 )
                painter.drawLine( p1, p2 )
                
                p1 = QPointF( r.center().x(), r.bottom() + 1 )
                p2 = QPointF( r.center().x(), r.bottom() + self.yDistance / 2 )
                painter.drawLine( p1, p2 )
                for i in range( 0, self.model().rowCount(index)):
                    r = self.itemRect( self.model().index( i, 0, index ) ).translated( offset.x(), offset.y() )
                    p1 = QPointF( r.center().x(), r.top() )
                    p2 = QPointF( r.center().x(), r.top() - self.yDistance / 2 )
                    painter.drawLine( p1, p2 )
                    
            elif rows == 1:
                p1 = QPointF( r.center().x(), r.bottom() + 1 )
                p2 = QPointF( p1.x(), p1.y() + self.yDistance - 2 )
                painter.drawLine( p1, p2 )
        painter.restore()
        
    def save(self, filename ):
        pix = QPixmap( self.realSize )
        pix.fill( Qt.transparent )
        painter = QPainter( pix )
        painter.setRenderHint( QPainter.Antialiasing )
        self.paintConnections( painter, QPointF( 0, 0 ) )
        self.paintItems(painter , QPointF( 0, 0 ) )
        painter.end()
        return pix.save( filename )

File: Trees/verticaltree.cpp
/*
 This file is part of Thistle.
    Thistle is free software: you can redistribute it and/or modify
    it under the terms of the Lesser GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License.
    Thistle is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    Lesser GNU General Public License for more details.
    You should have received a copy of the Lesser GNU General Public License
    along with Thistle.    If not, see <http://www.gnu.org/licenses/>.
 Thistle    Copyright (C) 2013    Dimitry Ernot & Romha Korev
*/
#include "verticaltree.h"
#include <QPen>
#include <QPainter>
#include <QPaintEvent>
#include <QScrollBar>
#include <QModelIndex>
#include <qmath.h>

namespace Thistle {

VerticalTree::ConnectorType VerticalTree::connectorType() const {
    const Q_D( VerticalTree );
    return (VerticalTree::ConnectorType)d->connectorType;
}


VerticalTree::VerticalTree( QWidget* parent ) : AbstractTree( parent ) {
    Q_D( VerticalTree );
    d->max = 0;
    d->min = 0;
    d->yDistance = 50;
    d->xDistance = 20;
    d->itemRect = QRect( -40, -20, 80, 40 );
    d->realSize = QSize( 100, 100 );
    d->connectorType = Straight;
    d->itemOffset = QPointF( 10, 10 );
}


void VerticalTree::paintConnectionsElbow( QPainter& painter, const QModelIndex& id ) const {
    const Q_D( VerticalTree );
    painter.save();
    painter.setPen( d->connectionPen );
    int rowCount = this->model()->rowCount( id );
    QRectF r = itemRect( id );
    if ( rowCount > 1 ) {
        QRectF child1 = itemRect( this->model()->index( 0, 0, id ) );
        QRectF child2 = itemRect( this->model()->index( rowCount - 1, 0, id ) );
        QPointF p1( child1.center().x(), child1.top() - d->yDistance / 2 );
        QPointF p2( child2.center().x(), child2.top() - d->yDistance / 2 );
        painter.drawLine( p1, p2 );
        p1 = QPointF( r.center().x(), r.bottom() + 1 );
        p2 = QPointF( r.center().x(), r.bottom() + d->yDistance / 2 );
        painter.drawLine( p1, p2 );
        for( int i = 0; i < this->model()->rowCount( id ); ++i ) {
            r = itemRect( this->model()->index( i, 0, id ) );
            p1 = QPointF( r.center().x(), r.top() );
            p2 = QPointF( r.center().x(), r.top() - d->yDistance / 2 );
            painter.drawLine( p1, p2 );
        }
    } else if ( rowCount == 1 ) {
        QPointF p1( r.center().x(), r.bottom() + 1 );
        QPointF p2( p1.x(), p1.y() + d->yDistance - 2 );
        painter.drawLine( p1, p2 );
    }
    painter.restore();
}


void VerticalTree::paintConnectionsFor( QPainter& painter, const QModelIndex& index, const QPointF& offset ) const {
    const Q_D( VerticalTree );
    painter.save();
    painter.setPen( d->connectionPen );
    if ( d->connectorType == Straight ) {
        AbstractTree::paintConnectionsFor( painter, index, offset );
    } else {
        int rows = this->model()->rowCount( index );
        QRectF r = itemRect(index).translated( offset.x(), offset.y() );
        if ( rows > 1 ) {
            QRectF child1 = itemRect( this->model()->index( 0, 0, index ) ).translated( offset.x(), offset.y() );
            QRectF child2 = itemRect( this->model()->index( rows - 1, 0, index ) ).translated( offset.x(), offset.y() );
            QPointF p1( child1.center().x(), child1.top() - d->yDistance / 2 );
            QPointF p2( child2.center().x(), child2.top() - d->yDistance / 2 );
            painter.drawLine( p1, p2 );
            p1 = QPointF( r.center().x(), r.bottom() + 1 );
            p2 = QPointF( r.center().x(), r.bottom() + d->yDistance / 2 );
            painter.drawLine( p1, p2 );
            for ( int i = 0; i < this->model()->rowCount(index); ++i ) {
                QRectF r = itemRect( this->model()->index( i, 0, index ) ).translated( offset.x(), offset.y() );
                p1 = QPointF( r.center().x(), r.top() );
                p2 = QPointF( r.center().x(), r.top() - d->yDistance / 2 );
                painter.drawLine( p1, p2 );
            }
        } else if ( rows == 1 ) {
                QPointF p1( r.center().x(), r.bottom() + 1 );
                QPointF p2( p1.x(), p1.y() + d->yDistance - 2 );
                painter.drawLine( p1, p2 );
        }
    }
    painter.restore();
}


void VerticalTree::paintConnectionsStraight( QPainter& painter, const QModelIndex& id ) const {
    const Q_D( VerticalTree );
    painter.save();
    painter.setPen( d->connectionPen );
    int rowCount = this->model()->rowCount( id );
    QPointF p1 = itemRect( id ).center();
    for( int i = 0; i < rowCount; ++i ) {
        QPointF p2 = itemRect( this->model()->index( i, 0, id ) ).center();
        painter.drawLine( p1, p2 );
    }
    painter.restore();
}
/***************************************
**
**                RESOLVING POSITIONS
**
***************************************/


void VerticalTree::positionsInTree() {
    Q_D( VerticalTree );
    d->itemTreePos.clear();
    d->depth = scan( this->model()->index(0,0), QPointF(0,0)).y();
    d->left = 0;
    //Q_FOREACH( QPointF p, myItemTreePos.values() ) {
    Q_FOREACH( QModelIndex idx, d->itemTreePos.keys() ) {
        QPointF p = d->itemTreePos[idx];
        d->left = qMax( d->left, p.x() );
    }
    d->depth -= 1;
    positionsInView();
}


void VerticalTree::positionsInView() {
    Q_D( VerticalTree );
    d->realSize.setWidth( ( d->left ) * ( d->xDistance + d->itemRect.width() ) + d->itemRect.width() );
    d->realSize.setHeight( d->depth * ( d->yDistance + d->itemRect.height() ) + d->itemRect.height() );
    setScrollBarValues();
    QPointF offset( d->itemRect.width()/2, d->itemRect.height()/2 );
    offset += d->itemOffset;
    Q_FOREACH( QModelIndex index, d->itemTreePos.keys() ) {
            QPointF p = d->itemTreePos[ index ];
            d->itemPos[index] = QPointF( p.x() * ( d->xDistance + d->itemRect.width() ), p.y() * ( d->yDistance + d->itemRect.height() ) ) + offset;
    }
}


QPointF VerticalTree::scan( const QModelIndex& index, const QPointF& ld ) {
    Q_D( VerticalTree );
    QPointF leftDepth( ld );
    int rows = this->model()->rowCount( index );
    if ( !index.isValid() ) {
        return QPointF(0, 0);
    } else if ( rows == 0 ) {
        setX( index, leftDepth.x() );
        setY(index, leftDepth.y());
        return QPointF(leftDepth.x() + 1, 1);
    }
    qreal childDepth = 0;
    for ( int r = 0; r < rows; ++r ) {
            QModelIndex child = index.child( r, 0 );
            QPointF p = scan( child, leftDepth + QPointF( 0, 1 ) );
            leftDepth.setX( p.x() );
            childDepth = qMax( childDepth, p.y() );
    }
    qreal left = d->itemTreePos[ this->model()->index(0, index.column(), index) ].x();
    qreal right = d->itemTreePos[ this->model()->index( rows - 1, index.column(), index ) ].x();
    if ( rows >= 2 ) {
            if ( rows % 2 == 1 ) {
                qreal r = qFloor( rows/2 ) + 1;
                qreal v = d->itemTreePos[ this->model()->index( r - 1, index.column(), index ) ].x();
                setX( index, v );
            } else {
                setX( index, (right + left) / 2 );
            }
    } else {
        setX( index, left );
    }
    setY(index, leftDepth.y());
    return QPointF( right + 1 , childDepth + 1 );
}


void VerticalTree::setConnectorType( VerticalTree::ConnectorType type ) {
    Q_D( VerticalTree );
    d->connectorType = type;
}


void VerticalTree::setScrollBarValues() {
    Q_D( VerticalTree );
    qreal dw = qMax( 0, ( (d->realSize.width() + d->itemRect.width() / 2 ) - width()) + 20);
    qreal dh = qMax( 0, (d->realSize.height() + d->itemRect.height()/2 - height()) + 20 );
    horizontalScrollBar()->setRange( 0, dw );
    verticalScrollBar()->setRange( 0, dh );
    d->itemOffset = QPointF( 10, 10 );
    qreal delta = width() - d->realSize.width();
    if ( delta > 20 ) {
        d->itemOffset = QPointF( delta/2.0, 10 );
    }
}


VerticalTree::~VerticalTree() {
}


}